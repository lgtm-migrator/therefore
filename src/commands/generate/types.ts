import type { TypescriptDefinition } from '../../definition'

export type ThereforeOutputType = 'jsonschema' | 'typescript' | 'validator'

export interface FileSymbol {
    symbolName: string
    schemaFile?: string
    compiledFile?: string
    definition: TypescriptDefinition
    typeOnly: boolean
}

export interface FileDefinition {
    file: string
    attachedFiles: { file: string; content: string; prettify: boolean; type: ThereforeOutputType }[]
    symbols: FileSymbol[]
    dependencies: Record<string, string[] | undefined>
}
