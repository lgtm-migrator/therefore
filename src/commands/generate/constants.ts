import { name } from '../../../package.json'

export const generatedBy = `Generated by ${name}`
