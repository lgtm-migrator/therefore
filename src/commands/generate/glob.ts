import { fixPath, fstat, isJust, isRight, Nothing } from '@zefiros-software/axioms'
import fastGlob from 'fast-glob'

import path from 'path'

// should give the same glob behaviour as Prettier
// https://github.com/prettier/prettier/blob/main/src/cli/expand-patterns.js
// also heavily inspired by the logic there :)
export async function expandGlobs({
    patterns,
    ignore = [],
    cwd,
    extension,
}: {
    patterns: string[]
    ignore?: string[]
    cwd: string
    extension: string
}): Promise<string[]> {
    const ignoredDirectories = ['.git', '.svn', '.hg', 'node_modules']
    const globOptions = {
        dot: true,
        ignore: [...ignoredDirectories.map((dir) => `**/${dir}`), ...ignore],
    }

    const entries: string[] = (
        await Promise.all(
            patterns.map(async (pattern) => {
                const absolutePath = path.resolve(cwd, pattern)
                if (ignoredDirectories.some((i) => pattern.includes(i))) {
                    return Nothing
                }

                const eitherStat = await fstat(absolutePath)
                if (isRight(eitherStat) && eitherStat.right !== Nothing) {
                    if (eitherStat.right.isFile()) {
                        return fastGlob.escapePath(fixPath(pattern))
                    } else if (eitherStat.right.isDirectory()) {
                        return `${fastGlob.escapePath(fixPath(path.relative(cwd, absolutePath) ?? '.'))}/**/*${extension}`
                    }

                    return Nothing
                } else if (pattern.startsWith('!')) {
                    globOptions.ignore.push(fixPath(pattern.slice(1)))
                    return Nothing
                }
                return fixPath(pattern)
            })
        )
    ).filter(isJust)
    return await fastGlob(entries, globOptions)
}
