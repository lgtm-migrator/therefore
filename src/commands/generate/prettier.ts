import type { ThereforeOutputType } from './types'

import type { Maybe } from '@zefiros-software/axioms'
import { Nothing } from '@zefiros-software/axioms'

export function maybeLoadPrettier() {
    // eslint-disable-next-line @typescript-eslint/consistent-type-imports
    let prettier: Maybe<typeof import('prettier')> = Nothing
    try {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        prettier = require('prettier')
    } catch {
        // prettier not found, skip formatting
    }
    return prettier
}

export async function formatFile(
    prettier: ReturnType<typeof maybeLoadPrettier>,
    input: string,
    file: string,
    type: ThereforeOutputType
): Promise<string> {
    if (prettier === Nothing) {
        return input
    }
    const options = await prettier.resolveConfig(file)
    return prettier.format(input, {
        ...(options ?? {}),
        ...(type === 'jsonschema' ? { parser: 'json' } : {}),
    })
}
