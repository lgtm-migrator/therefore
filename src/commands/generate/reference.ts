import type { FileDefinition } from './types'

import type { TypescriptReference, TypescriptDefinition } from '../../definition'
import type { ThereforeCst } from '../../lib/types/types'
import { toTypescriptDefinition } from '../../lib/visitor'

import { evaluate } from '@zefiros-software/axioms'

interface ExpandedReference {
    file: string
    symbolName: string
    uuid: string
    referenceName: string
}

export function requireReference(
    definitions: Record<string, FileDefinition>,
    current: FileDefinition,
    ref: TypescriptReference,
    locals: NonNullable<TypescriptDefinition['locals']>
): ExpandedReference[] {
    const found = Object.values(definitions)
        .flatMap((d) =>
            d.symbols.map((s) => ({
                uuid: s.definition.uuid,
                symbolName: s.definition.symbolName,
                referenceName: s.definition.referenceName,
                file: d.file,
            }))
        )
        .find((s) => s.uuid === ref.uuid)

    if (!found) {
        // this is a locally defined variable
        const [refName, reference] = ref.reference
        const sourceSymbol = `${refName}Local`
        const definition = toTypescriptDefinition({
            sourceSymbol,
            schema: evaluate(reference) as ThereforeCst,
            exportSymbol: false,
            locals,
        })
        current.symbols.push({
            definition,
            symbolName: sourceSymbol,
            typeOnly: true,
        })

        // make the DFS well defined
        const extra = definition.references.map((r) => requireReference(definitions, current, r, locals)).flat()
        return [
            ...extra,
            {
                file: current.file,
                symbolName: definition.symbolName,
                referenceName: definition.referenceName,
                uuid: ref.uuid,
            },
        ]
    }
    return [
        {
            file: found.file,
            symbolName: found.symbolName,
            referenceName: found.referenceName,
            uuid: ref.uuid,
        },
    ]
}
