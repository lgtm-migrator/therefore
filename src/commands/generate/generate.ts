import { generatedBy } from './constants'
import { expandGlobs } from './glob'
import { formatFile, maybeLoadPrettier } from './prettier'
import { requireReference } from './reference'
import type { FileDefinition, ThereforeOutputType } from './types'

import { version } from '../../../package.json'
import { renderTemplate } from '../../common/template/template'
import type { TypescriptDefinition } from '../../definition'
import type { CstNode } from '../../lib/cst/cst'
import { isThereforeExport } from '../../lib/guard'
import type { ThereforeCst } from '../../lib/types/types'
import { prepass, toJsonSchema } from '../../lib/visitor'
import { toTypescriptDefinition, writeThereforeSchema } from '../../lib/visitor/typescript/typescript'

import { evaluate, hasPropertiesDefined, isDefined } from '@zefiros-software/axioms'
import decamelize from 'decamelize'

import fs from 'fs'
import path from 'path'

function requireModule(module: string): Record<string, CstNode | unknown> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/').replace('.ts', '')

    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return require(relative.startsWith('.') ? relative : `./${relative}`) as Record<string, CstNode | unknown>
}

export function scanModule({
    entry,
    fullPath,
    compile,
    require = requireModule,
}: {
    entry: string
    fullPath: string
    compile: boolean
    require?: (module: string) => Record<string, CstNode | unknown>
}): FileDefinition {
    const file: FileDefinition = {
        file: fullPath,
        attachedFiles: [],
        symbols: [],
        dependencies: {},
    }
    const module = require(entry)

    for (const [sourceSymbol, symbol] of Object.entries(module).filter(([, exports]) => isThereforeExport(exports))) {
        const simplified = prepass(symbol as ThereforeCst)
        const schemaName = decamelize(sourceSymbol, { separator: '-' })
        const schemaFile = `./schemas/${schemaName}.schema.json`
        const compiledFile = `./schemas/${schemaName}.schema.js`

        const jsonschema = toJsonSchema(simplified, compile)
        const definition = toTypescriptDefinition({ sourceSymbol, schema: simplified })

        file.symbols.push(
            ...Object.values(definition.locals ?? {}).map((local) => ({
                symbolName: local.symbolName,
                definition: local,
                typeOnly: true,
            }))
        )

        file.symbols.push({
            symbolName: sourceSymbol,
            definition: definition,
            schemaFile,
            compiledFile,
            typeOnly: false,
        })

        if (definition.isExported) {
            if (jsonschema.compiled === true) {
                const filePath = path.join(path.dirname(entry), compiledFile)
                file.attachedFiles.push({
                    file: filePath,
                    content: `/**\n * ${generatedBy}\n * eslint-disable\n */\n${jsonschema.code}`,
                    prettify: false,
                    type: 'validator',
                })
            } else {
                file.attachedFiles.push({
                    file: path.join(path.dirname(entry), schemaFile),
                    content: JSON.stringify(jsonschema.schema, null, 2),
                    prettify: true,
                    type: 'jsonschema',
                })
            }

            console.debug(` - found ${definition.symbolName}`)
        }
    }
    return file
}

export function scanFiles({
    files,
    basePath,
    compile,
}: {
    files: string[]
    basePath: string
    compile: boolean
}): Record<string, FileDefinition> {
    const definitions: Record<string, FileDefinition> = {}
    for (const entry of files) {
        const fullPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = fullPath.substr(0, fullPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)
        try {
            const file = scanModule({ entry, fullPath, compile })
            if (file.symbols.length > 0) {
                definitions[fullPath] = file
            }
        } catch (e: unknown) {
            const error = e as Error
            console.debug(error.message, error.stack)
            throw error
        }
    }
    return definitions
}

// eslint-disable-next-line @typescript-eslint/require-await
export async function compileSchemas(
    entries: string[],
    { compile = false, cwd, outputFileRename }: { cwd: string; compile: boolean; outputFileRename: (path: string) => string }
): Promise<
    {
        file: string
        template: string
        data: Record<string, string>
        type: ThereforeOutputType
        prettify: boolean
    }[]
> {
    const schemaFiles: {
        file: string
        relativeSource: string
        template: string
        data: Record<string, string>
        type: ThereforeOutputType
        prettify: boolean
    }[] = []
    const definitions = scanFiles({ files: entries, basePath: cwd, compile })

    const localReferences = {}

    for (const [file, definition] of Object.entries(definitions)) {
        const locals: NonNullable<TypescriptDefinition['locals']> = {}
        const required = definition.symbols
            .flatMap((i) => i.definition.references.map((r) => requireReference(definitions, definition, r, locals)))
            .flat(2)

        definition.symbols.push(
            ...Object.values(locals ?? {}).map((local) => ({
                symbolName: local.sourceSymbol,
                definition: local,
                typeOnly: true,
            }))
        )

        for (const r of required) {
            definition.dependencies[r.file] ??= []
            definition.dependencies[r.file]?.push(r.symbolName)
        }

        const schemaFileName = outputFileRename(file)

        const relativeDependencies = Object.entries(definition.dependencies)
            .map(([f, deps]) => ({
                file: outputFileRename(f),
                deps: `{ ${[...new Set(deps)].sort().join(', ')} }`,
            }))
            .filter(({ file: f }) => f !== schemaFileName)
            .map(({ file: f, deps }) => {
                const otherPath = path.relative(path.dirname(schemaFileName), f).replace('.ts', '').replace(/\\/g, '/')
                return `import ${deps} from '${otherPath.startsWith('.') ? otherPath : `./${otherPath}`}'`
            })

        const references = Object.fromEntries(required.map((r) => [r.uuid, r.referenceName] as const))

        const schemaFile = path.relative(cwd, schemaFileName).replace(/\\/g, '/')
        const relativeSource = `./${path.relative(path.dirname(schemaFile), path.relative(cwd, file)).replace(/\.ts$/g, '')}`
        schemaFiles.push(
            {
                file: schemaFile,
                relativeSource,
                type: 'typescript',
                template: ([] as string[])
                    .concat(
                        ['/**', ` * ${generatedBy}@v${version}`, ' * Do not manually touch this', ' */\n'],
                        ['/* eslint-disable */'],
                        definition.symbols
                            .filter(hasPropertiesDefined(['schemaFile']))
                            .filter((i) => !i.typeOnly && i.compiledFile === undefined)
                            .map((i) => `import ${i.symbolName}Schema from '${i.schemaFile}'`)
                            .sort(),
                        relativeDependencies.length ? ['', ...relativeDependencies] : [],
                        ['', `import type { ValidateFunction } from 'ajv'`, ''],
                        ['', `import AjvValidator from 'ajv'`, ''],
                        Object.values(definition.symbols)
                            .map((i) => [
                                i.definition.declaration,
                                i.definition.isExported && !i.typeOnly
                                    ? writeThereforeSchema({
                                          symbolName: i.definition.referenceName,
                                          schemaReference: `${i.symbolName}Schema`,
                                          validatorFile: i.compiledFile,
                                          source: relativeSource.replace('./', ''),
                                          sourceSymbol: i.definition.sourceSymbol,
                                          description: evaluate(i.definition.schema).description,
                                      })
                                    : undefined,
                            ])
                            .flat(2)
                            .filter(isDefined)
                    )
                    .join('\n'),
                data: { ...localReferences, ...references },
                prettify: true,
            },
            ...definition.attachedFiles.map((j) => ({
                file: path.relative(cwd, j.file).replace(/\\/g, '/'),
                relativeSource,
                type: j.type,
                template: j.content,
                data: { ...localReferences, ...references },
                prettify: j.prettify,
            }))
        )
    }

    return compile ? schemaFiles : schemaFiles.filter((f) => f.type !== 'validator')
}

export async function generate({
    globs,
    ignore,
    extension,
    compile,
    outputFileRename,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    compile: boolean
    outputFileRename: (path: string) => string
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const schemaFiles = await compileSchemas(entries, {
        cwd,
        outputFileRename,
        compile,
    })
    const prettier = maybeLoadPrettier()

    for (const { file, template, data, type, prettify } of schemaFiles) {
        fs.mkdirSync(path.dirname(file), { recursive: true })
        const contents = renderTemplate(template, data)
        fs.writeFileSync(file, prettify ? await formatFile(prettier, contents, file, type) : contents)
    }
}
