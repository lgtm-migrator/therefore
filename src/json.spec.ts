import { compileSchemas } from './commands/generate/generate'

test('json', async () => {
    expect(
        await compileSchemas(['examples/json/json.schema.ts'], {
            outputFileRename: (file: string) => file.replace('.ts', '.type.ts'),
            cwd: process.cwd(),
            compile: false,
        })
    ).toMatchSnapshot()
})

test('simple', async () => {
    expect(
        await compileSchemas(['examples/json/simple.schema.ts'], {
            outputFileRename: (file: string) => file.replace('.ts', '.type.ts'),
            cwd: process.cwd(),
            compile: false,
        })
    ).toMatchSnapshot()
})
