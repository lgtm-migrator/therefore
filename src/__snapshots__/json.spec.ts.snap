// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json 1`] = `
Array [
  Object {
    "data": Object {
      "00010-000": "JsonAdv",
      "00019-000": "JsonRefLocal",
      "0009-000": "Json",
    },
    "file": "examples/json/json.schema.type.ts",
    "prettify": true,
    "relativeSource": "./json.schema",
    "template": "/**
 * Generated by @zefiros-software/therefore@v0.0.1
 * Do not manually touch this
 */

/* eslint-disable */

import type { ValidateFunction } from 'ajv'


import AjvValidator from 'ajv'

export type Json = string | null | boolean | number | {
    [k: string]: ( {{0009-000}} ) | undefined
} | ({{0009-000}})[]

export const Json = {
    validate: require('./schemas/json.schema.js') as ValidateFunction<Json>,
    get schema() { return Json.validate.schema},
    source: \`\${__dirname}json.schema\`,
    sourceSymbol: 'json',
    is: (o: unknown): o is Json => Json.validate(o) === true,
    assert: (o: unknown): asserts o is Json => {
        if (!Json.validate(o)) {
            throw new AjvValidator.ValidationError(Json.validate.errors ?? [])
        }
    },
} as const

export type JsonAdv = {{00019-000}}

export const JsonAdv = {
    validate: require('./schemas/json-adv.schema.js') as ValidateFunction<JsonAdv>,
    get schema() { return JsonAdv.validate.schema},
    source: \`\${__dirname}json.schema\`,
    sourceSymbol: 'jsonAdv',
    is: (o: unknown): o is JsonAdv => JsonAdv.validate(o) === true,
    assert: (o: unknown): asserts o is JsonAdv => {
        if (!JsonAdv.validate(o)) {
            throw new AjvValidator.ValidationError(JsonAdv.validate.errors ?? [])
        }
    },
} as const

type JsonRefLocal = string | null | boolean | number | {
    [k: string]: ( {{00010-000}} ) | undefined
} | ({{00010-000}})[]
",
    "type": "typescript",
  },
  Object {
    "data": Object {
      "00010-000": "JsonAdv",
      "00019-000": "JsonRefLocal",
      "0009-000": "Json",
    },
    "file": "examples/json/schemas/json.schema.json",
    "prettify": true,
    "relativeSource": "./json.schema",
    "template": "{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"oneOf\\": [
    {
      \\"type\\": \\"string\\"
    },
    {
      \\"type\\": \\"null\\"
    },
    {
      \\"type\\": \\"boolean\\"
    },
    {
      \\"type\\": \\"number\\"
    },
    {
      \\"type\\": \\"object\\",
      \\"additionalProperties\\": {
        \\"$ref\\": \\"#\\"
      }
    },
    {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"$ref\\": \\"#\\"
      }
    }
  ]
}",
    "type": "jsonschema",
  },
  Object {
    "data": Object {
      "00010-000": "JsonAdv",
      "00019-000": "JsonRefLocal",
      "0009-000": "Json",
    },
    "file": "examples/json/schemas/json-adv.schema.json",
    "prettify": true,
    "relativeSource": "./json.schema",
    "template": "{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"$ref\\": \\"#/definitions/{{00019-000}}\\",
  \\"definitions\\": {
    \\"{{00019-000}}\\": {
      \\"oneOf\\": [
        {
          \\"type\\": \\"string\\"
        },
        {
          \\"type\\": \\"null\\"
        },
        {
          \\"type\\": \\"boolean\\"
        },
        {
          \\"type\\": \\"number\\"
        },
        {
          \\"type\\": \\"object\\",
          \\"additionalProperties\\": {
            \\"$ref\\": \\"#\\"
          }
        },
        {
          \\"type\\": \\"array\\",
          \\"items\\": {
            \\"$ref\\": \\"#\\"
          }
        }
      ]
    }
  }
}",
    "type": "jsonschema",
  },
]
`;

exports[`simple 1`] = `
Array [
  Object {
    "data": Object {
      "0004-000": "Simple",
    },
    "file": "examples/json/simple.schema.type.ts",
    "prettify": true,
    "relativeSource": "./simple.schema",
    "template": "/**
 * Generated by @zefiros-software/therefore@v0.0.1
 * Do not manually touch this
 */

/* eslint-disable */

import type { ValidateFunction } from 'ajv'


import AjvValidator from 'ajv'

export type Simple = number | ({{0004-000}})[]

export const Simple = {
    validate: require('./schemas/simple.schema.js') as ValidateFunction<Simple>,
    get schema() { return Simple.validate.schema},
    source: \`\${__dirname}simple.schema\`,
    sourceSymbol: 'simple',
    is: (o: unknown): o is Simple => Simple.validate(o) === true,
    assert: (o: unknown): asserts o is Simple => {
        if (!Simple.validate(o)) {
            throw new AjvValidator.ValidationError(Simple.validate.errors ?? [])
        }
    },
} as const
",
    "type": "typescript",
  },
  Object {
    "data": Object {
      "0004-000": "Simple",
    },
    "file": "examples/json/schemas/simple.schema.json",
    "prettify": true,
    "relativeSource": "./simple.schema",
    "template": "{
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\",
  \\"oneOf\\": [
    {
      \\"type\\": \\"number\\"
    },
    {
      \\"type\\": \\"array\\",
      \\"items\\": {
        \\"$ref\\": \\"#\\"
      }
    }
  ]
}",
    "type": "jsonschema",
  },
]
`;
